src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector.cs

```diff
@@ -18,77 +16,76 @@ namespace Nethermind.Consensus.Processing;
-            );
-    private const int CacheSize = 4;
-    private readonly LruCache<long, bool>
-        _censorshipDetector = new(CacheSize, CacheSize, "censorshipDetector");
+    private readonly LruCache<long, bool> _temporaryCensorshipDetector;
+    private readonly LruCache<long, bool> _censorshipDetector;
+    private const int _temporaryCacheSize = 16;
+    private const int _cacheSize = 4;
 
     public CensorshipDetector(
         ITxPool txPool,
+        IComparer<Transaction> comparer,
         IBlockProcessor blockProcessor,
         IBlockTree blockTree,
-        IAccountStateProvider accounts,
         ILogManager logManager)
     {
         _txPool = txPool;
+        _comparer = comparer;
         _blockProcessor = blockProcessor;
         _blockTree = blockTree;
-        _accounts = accounts;
         _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+        _temporaryCensorshipDetector = new(
+            _temporaryCacheSize, _temporaryCacheSize, "temporaryCensorshipDetector"
+            );
+        _censorshipDetector = new(_cacheSize, _cacheSize, "censorshipDetector");
         _blockProcessor.BlockProcessing += OnBlockProcessing;
-        _blockTree.BlockAddedToMain += OnBlockAdded;
+        _blockTree.BlockAddedToMain += OnBlockAddedToMain;
     }
 
     private void OnBlockProcessing(object? sender, BlockEventArgs e)
     {
         Task.Run(() => PreCachingOperation(e.Block));
     }
 
-    private void OnBlockAdded(object? sender, BlockEventArgs e)
+    private void OnBlockAddedToMain(object? sender, BlockReplacementEventArgs e)
     {
         Task.Run(() => Cache(e.Block));
     }
 
     private void PreCachingOperation(Block block)
     {
-        UInt256 maxGasPriceInBlock = 0;
-        Transaction[] blockTransactions = block.Transactions;
-        foreach (Transaction tx in blockTransactions)
+        Transaction bestTxInBlock = new Transaction();
+
+        foreach (Transaction tx in block.Transactions)
         {
-            if (!tx.SupportsBlobs)
+            if (!tx.SupportsBlobs && _comparer.Compare(bestTxInBlock, tx) > 0)
             {
-                if (tx.Supports1559)
-                {
-                    maxGasPriceInBlock = UInt256.Max(maxGasPriceInBlock, tx.GasPrice);
-                }
-                else
-                {
-                    maxGasPriceInBlock = UInt256.Max(maxGasPriceInBlock, tx.GasPrice + block.BaseFeePerGas);
-                }
+                bestTxInBlock = tx;
             }
         }
 
-        Transaction bestTx = _txPool.GetBestTxOfEachSender().First();
-        UInt256 maxGasPriceInPool = bestTx.Supports1559 ? bestTx.GasPrice + block.BaseFeePerGas : bestTx.GasPrice;
+        Transaction bestTxInPool = _txPool.GetBestTx();
+
+        bool unconfirmedBlockPotentialCensorship = _comparer.Compare(bestTxInBlock, bestTxInPool) > 0;
```

Comment:
and then do it only if `bestTxInPool` is not `null`. But not sure if worth refactoring